### PR TITLE
NCS 86 company officer validation

### PR DESCRIPTION
**Pull request by LewisJohnEvans**
Company officer validation currently unreachable. Can be added to list of other eligibility rules for validation in later commit when ready.

### CODE REVIEW

File: src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidationTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        CompanyOfficerApi director = om.readValue(SECRETARY_JSON,CompanyOfficerApi.class);
+        CompanyOfficerApi secretary = om.readValue(DIRECTOR_JSON,CompanyOfficerApi.class);
+        OFFICER_LIST.add(director);
+        OFFICER_LIST.add(secretary);
+        mockOfficers.setItems(OFFICER_LIST);
+        mockOfficers.setActiveCount((long) OFFICER_LIST.size());
+
+        var result = companyOfficerValidation.getOfficerCount(mockOfficers.getItems(), mockOfficers.getActiveCount());
+        assertEquals(result, 2L);
+        assertNotEquals(result, OFFICER_LIST.size());
```
Comment (ICoxCH) [inline src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidationTest.java:100]: these two have their arguments the wrong way around.
It should be `assert(expected, actual)` so in this case `assertEquals(2L, result)`

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/EligibilityFailureReason.java [hunk 2]
```diff
@@ -5,5 +5,6 @@
     INVALID_COMPANY_STATUS,
     INVALID_COMPANY_TYPE_CS01_FILING_NOT_REQUIRED,
     INVALID_COMPANY_TYPE_PAPER_FILING_ONLY,
-    INVALID_COMPANY_TYPE_USE_WEB_FILING
+    INVALID_COMPANY_TYPE_USE_WEB_FILING,
+    INVALID_OFFICER_COUNT
```
Comment (ICoxCH) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/EligibilityFailureReason.java:9]: Similar to this comment [here](https://github.com/companieshouse/confirmation-statement-api/pull/27#discussion_r637736958)
I'd reccommend being more descriptive. Perhaps: `INVALID_COMPANY_APPOINTMENTS_MORE_THAN_ONE_OFFICER`
So a user only interacting with the api has more of an idea what is wrong with the officer count

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidation.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        var officers = officerService.getOfficers(companyProfileApi.getCompanyNumber());
+        var officerCount = getOfficerCount(officers.getItems(), officers.getActiveCount());
+        if (officerCount != null && officerCount > 1) {
+            throw new EligibilityException(EligibilityFailureReason.INVALID_OFFICER_COUNT);
+        }
+    }
+
+    public Long getOfficerCount(List<CompanyOfficerApi> officers, Long activeCount) {
+        for(CompanyOfficerApi i: officers) {
+            if (i.getOfficerRole().getOfficerRole().contains("secretary")) activeCount--;
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidation.java:34]: CH java guidelines recommend following the google guidelines, a single line will need to be wrapped with curly braces - CH - https://github.com/companieshouse/styleguides/blob/master/java.md
google -  https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used

File: src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidationTest.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import uk.gov.companieshouse.api.model.officers.OfficersApi;
+import uk.gov.companieshouse.confirmationstatementapi.eligibility.EligibilityFailureReason;
+import uk.gov.companieshouse.confirmationstatementapi.exception.EligibilityException;
+import uk.gov.companieshouse.confirmationstatementapi.exception.ServiceException;
+import uk.gov.companieshouse.confirmationstatementapi.service.OfficerService;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidationTest.java:22]: shouldn't have the import *, need to import each class separately

File: src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidationTest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    @Mock
+    private OfficerService officerService;
+
+    @InjectMocks
+    private CompanyOfficerValidation companyOfficerValidation;
+
+    @BeforeEach
+    void init() throws JsonProcessingException {
+        OFFICER_LIST.clear();
+        CompanyOfficerApi MOCK_OFFICER = om.readValue(DIRECTOR_JSON, CompanyOfficerApi.class);
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidationTest.java:47]: you don't need to use the ObjectMapper here, you can just create the MOCK_OFFICER with new(). eg.
CompanyOfficerApi MOCK_OFFICER = new CompanyOfficerApi();
        MOCK_OFFICER.setOfficerRole(OfficerRoleApi.DIRECTOR);

File: src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidationTest.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        mockOfficers.setActiveCount((long) OFFICER_LIST.size());
+
+        when(officerService.getOfficers(COMPANY_NUMBER)).thenReturn(mockOfficers);
+
+       assertDoesNotThrow(() -> companyOfficerValidation.validate(companyProfileApi));
+    }
+
+    @Test
+    void validateThrowsOnMultipleOfficerCompany() throws JsonProcessingException, ServiceException {
+        CompanyOfficerApi director = om.readValue(DIRECTOR_JSON,CompanyOfficerApi.class);
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyOfficerValidationTest.java:64]: same here, I would just use
CompanyOfficerApi director = new CompanyOfficerApi();
        director.setOfficerRole(OfficerRoleApi.DIRECTOR);

or - make a new private function createDirectorOfficer() and put those 2 lines above in there. Then call it when you need a new director.

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/EligibilityFailureReason.java (hunk 2 above)
Comment (LewisJohnEvans) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/EligibilityFailureReason.java:9]: I took INVALID_OFFICER_COUNT from the Validation errors enum in the schema.json shall i replace that with the more descriptive version or add this as a new entry and leave that alone ?

Comment (ICoxCH) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/EligibilityFailureReason.java:9]: yes best to replace it with the more descriptive one.
I added INVALID_OFFICER_COUNT way back when there was one type of validation on officers. Now with multiple types, i don't think we'd ever use the more generic "INALID_OFFICER_COUNT"

**Review by ICoxCH — APPROVED** [review]

**Pull request merged**